### PR TITLE
Dropped support for per_page pagination

### PR DESCRIPTION
**Pull request by ShimShtein**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -96,9 +82,7 @@ def process_index(base, serializer_class, opts: {}, default_sort: nil, allowed_s
 
   def index_meta(scope)
     meta = {
-      :total => scope.count,
-      :per_page => scope.per_page,
-      :page => scope.current_page
+      :total => scope.limit(nil).offset(nil).count
```
Comment (adamruzicka) [inline app/controllers/application_controller.rb]: Do we really need to unset the limit and offset to get the correct count?

Comment (ShimShtein) [reply app/controllers/application_controller.rb]: Yep:

```
[3] pry(main)> App.offset(5).offset(nil).count
   (2.1ms)  SELECT COUNT(*) FROM "apps"
=> 10
[4] pry(main)> App.offset(5).count
   (0.8ms)  SELECT COUNT(*) FROM (SELECT 1 AS one FROM "apps" OFFSET $1) subquery_for_count  [["OFFSET", 5]]
=> 5
```

The same goes for limit.

**Comment by ShimShtein** [comment]
@adamruzicka, now with links. Should be ready for review.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -29,20 +30,6 @@ class BadRequest < RuntimeError; end
   end
 
   def paginate(scope)
-    raise BadRequest, 'Both page and offset pagination is not allowed' if page_pagination? && offset_pagination?
-
-    limit_query(scope.paginate(:per_page => params[:per_page] || 10, :page => params[:page] || 1))
```
Comment (adamruzicka) [inline app/controllers/application_controller.rb]: I'd keep the behaviour similar to the original pagination and provide default values for limit and offset

**Comment by ShimShtein** [comment]
Added default limit and changed next and last to pages

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -90,21 +73,32 @@ def process_index(base, serializer_class, opts: {}, default_sort: nil, allowed_s
...
+    meta[:limit] = scope.limit_value if scope.respond_to?(:limit_value) && scope.limit_value
+    meta[:offset] = scope.offset_value if scope.respond_to?(:offset_value) && scope.offset_value
     meta
   end
 
+  def index_links(_scope, total_records)
+    links = {
+      first: same_request_with_parameters(first_records_parameters),
+      last: same_request_with_parameters(last_records_parameters(total_records)),
+      next: same_request_with_parameters(next_records_parameters)
```
Comment (adamruzicka) [inline app/controllers/application_controller.rb]: `next` should be omitted (or `nil`) the same way `previous` is if there are no more records

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -102,35 +52,42 @@ def process_index(base, serializer_class, opts: {}, default_sort: nil, allowed_s
...
 
-  def render_not_found(what, id)
-    message = "Could not find #{what}"
-    message += "with 'id'=#{id}" if id
-    render :json => { :errors => message }, :status => :not_found
+    offset = params[:offset]
+    if offset.to_i + limit_param < total_records
+      links[:next] = same_request_with_parameters(next_records_parameters)
+    end
+    links[:previous] = same_request_with_parameters(previous_records_parameters) if offset
```
Comment (adamruzicka) [inline app/controllers/application_controller.rb]: Requesting a page with `limit=10` and `offset=0` leads to `previous` being `http://localhost:3000/api/hooks/endpoints?limit=0&offset=0` which isn't particularly useful

**Comment by adamruzicka** [comment]
Thanks @ShimShtein !

**Pull request merged**